### ISSUE TITLE
Feature/An API that fetches information about applications installed on an OS

### ISSUE THREAD

**Issue by SaeedZhiany** [oshi/oshi#2581]
Using this library to fetch information about installed applications on the host operating system would be nice.

Each application's information will contain

- application name
- installed date
- version
- manufacture (or provider or producer company name)
- OS Compatibility (x86 / x64)
- etc

like below (the picture is taken from the Application IOBitUninstaller)

<img width="869" alt="image" src="https://github.com/oshi/oshi/assets/13766432/f0a23400-9317-4185-a093-6e8187602c83">

**Comment by dbwiddis** [comment]
Hey @SaeedZhiany thanks for the suggestion.

If you want to try to submit a PR implementing a cross-platform feature to do this, you're welcome to do so.  It's unlikely one of the current maintainers will pick this up.

We did do something similar but with a more limited scope implementing the "Services" API, and even that proved a challenge.  Some questions you'll need to answer:
 - what is the definition of an "application"?  
   - On macOS I can look in the Applications folder but I'll miss other applications that install themselves in other directories.  How do I search the entire system (quickly) to find these?
   - On WIndows does every .exe file count?
   - What counts as an "application" on command-line based systems?  Or are we only really dealing with UI-based systems?
   - how does this differ from checking for installed "packages" with a package manager?
 - Does every file systemwide with the executable bit on it count on *nix systems?  How do I search the entire system (quickly) to find these?

It looks like there could be some alternatives:
 - Windows apparently has "applications" In the registry: https://superuser.com/questions/1603763/how-can-i-run-a-single-command-to-show-all-installed-applications-in-windows-10
 - macOS has them somewhere that can be fetched (command line `system_profiler SPApplicationsDataType` but may be a native alternative (or maybe not): https://apple.stackexchange.com/questions/98826/how-to-get-a-complete-list-of-all-applications-installed 
 - For Linux/Unix, it looks like one would just try all the different package managers: https://unix.stackexchange.com/questions/20979/how-do-i-list-all-installed-programs (and perhaps for macOS one might try homebrew or macports lists as well)

Feel free to contribute a PR.

**Comment by dbwiddis** [comment]
> I'm currently planning to support only Windows and Linux

At least include macOS. The system profiler command line I quoted shouldn't be too hard to parse.

For Linux, see the link I gave, there are many more package managers.

**Comment by rohan-coder02** [comment]
Hi @dbwiddis 

Below is my analysis and code changes to feature an API to fetch Installed Applications across OS platforms (Mac, Windows, and Linux)

**Command Output Findings:**

- Each OS has different variables in the command output i.e. we cannot use the same DTO for all
- All variables are considered in the DTO for Mac (only 8 variables in the output) 
- In the case of Linux and Windows, the command outputs has more variables, hence, have taken into account the important 6-7 variables in Linux and Windows 
- Additional variables can be added in the future for DTOs of Linux and Windows based on requirements/ask

_Mac:_ 

The following command helps get installed apps data -> `system_profiler SPApplicationsDataType`
(All variables are considered in the DTO)

_Windows:_

The following commands help to get installed apps 

reg query "HKLM\Software\Microsoft\Windows\CurrentVersion\Uninstall" /s
reg query "HKLM\Software\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall" /s
reg query "HKCU\Software\Microsoft\Windows\CurrentVersion\Uninstall" /s

_Linux:_

Debian -> `dpkg-query -W -f='${Package}|${Version}|${Architecture}|${Installed-Size}|${Maintainer}|${Source}|${Homepage}\n'`
Fedora ->` rpm -qa --queryformat '%{NAME}|%{VERSION}-%{RELEASE}|%{ARCH}|%{SIZE}|%{PACKAGER}|%{SOURCERPM}|%{URL}\n'`

In the output of `cat /etc/os-release`, I'm checking the ID_LIKE variable is fedora or debian
This approach covers the popular flavors of linux such as Ubuntu, Redhat, CentOS, Amazon Linux 2, etc
We can further add commands for other linux flavors

**Code Changes Implemented in OSHI Codebase:**

In **oshi.software.os** added interface InstalledApps

This consists of getInstalledApps function and DTOs for Linux, Mac, and Windows

```
public interface InstalledApps {

    List<?> getInstalledApps();

    final class LinuxAppInfo {
        private String packageName;
        private String version;
        private String architecture;
        private String installedSize;
        private String maintainer;
        private String source;
        private String homepage;

       //constructor
       //getter for above variables
       //toString
     }
     
     final class MacAppInfo {
        private String appName;
        private String version;
        private String obtainedFrom;
        private String lastModified;
        private String kind;
        private String signedBy;
        private String location;
        private String getInfoString;

        //constructor
       //getter for above variables
       //toString
     }

    final class WindowsAppInfo {

        private final String name;
        private final String version;
        private final String vendor;
        private final String installDate;
        private final String installLocation;
        private final String installSource;
      
        //constructor
       //getter for above variables
       //toString
     }
}
```

In **oshi.software.common** added an abstract class AbstractInstalledApps

```
abstract public class AbstractInstalledApps implements InstalledApps {
    public abstract List<?> getInstalledApps();
}
```

_Mac:_

```
In oshi.software.os.mac

public class MacInstalledApps extends AbstractInstalledApps {

    private final String colon = ":";

    @Override
    public List<MacAppInfo> getInstalledApps() {
        List<String> macOutput = ExecutingCommand.runNative("system_profiler SPApplicationsDataType");
        return parseMacAppInfo(macOutput);
    }
    
    private List<MacAppInfo> parseMacAppInfo(List<String> lines) {
            //logic to parse the output and store it in a list of MacAppInfo DTO
    }
```

_Windows:_

```
In oshi.driver.windows.registry

public class InstalledAppsData {
    private InstalledAppsData() {}

    private static final List<String> UNINSTALL_PATHS = Arrays.asList(
        "HKLM\\Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall",
        "HKLM\\Software\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall",
        "HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall"
    );

    public enum AppInfo {
        NAME, VERSION, VENDOR, INSTALLDATE, INSTALLLOCATION, INSTALLSOURCE;
    }

    public static List<Map<AppInfo, String>> queryInstalledApps() {
        List<Map<AppInfo, String>> appInfoList = new ArrayList<>();

        for (String path : UNINSTALL_PATHS) {
            List<String> output = ExecutingCommand.runNative("reg query " + path + " /s");
            parseRegistryOutput(output, appInfoList);
        }

        return appInfoList;
    }

    private static void parseRegistryOutput(List<String> output, List<Map<AppInfo, String>> appInfoList) {
         //logic to parse the output and store it in a list of WindowsAppInfo DTO
    }
}


In oshi.software.os.windows

public class WindowsInstalledApps extends AbstractInstalledApps {

    public List<WindowsAppInfo> getInstalledApps() {
    List<WindowsAppInfo> appInfoList = new ArrayList<>();

    List<Map<InstalledAppsData.AppInfo, String>> registryResults = InstalledAppsData.queryInstalledApps();

    for (Map<InstalledAppsData.AppInfo, String> app : registryResults) {
        appInfoList.add(new WindowsAppInfo(app.getOrDefault(InstalledAppsData.AppInfo.NAME, ""),
                app.getOrDefault(InstalledAppsData.AppInfo.VERSION, ""),
                app.getOrDefault(InstalledAppsData.AppInfo.VENDOR, ""),
                app.getOrDefault(InstalledAppsData.AppInfo.INSTALLDATE, ""),
                app.getOrDefault(InstalledAppsData.AppInfo.INSTALLLOCATION, ""),
                app.getOrDefault(InstalledAppsData.AppInfo.INSTALLSOURCE, "")));
    }

    return appInfoList;
}

}
```

_Linux:_

```
In oshi.software.os.linux

public class LinuxInstalledApps extends AbstractInstalledApps {

   private static final Logger LOG = LoggerFactory.getLogger(LinuxInstalledApps.class);

    private enum LinuxFlavor {
        DEBIAN, RPM, UNKNOWN
    }

    @Override
    public List<LinuxAppInfo> getInstalledApps() {
        LinuxFlavor flavor = detectLinuxFlavor();
        List<String> output = fetchInstalledApps(flavor);
        return parseApplicationInfo(output);
    }

    private LinuxFlavor detectLinuxFlavor() {
        byte[] fileBytes = FileUtil.readAllBytes("/etc/os-release", true);
        List<String> lines = ParseUtil.parseByteArrayToStrings(fileBytes);

        for (String line : lines) {
            if (line.startsWith("ID_LIKE")) {
                if (line.contains("debian")) {
                    return LinuxFlavor.DEBIAN;
                } else if (line.contains("fedora")) {
                    return LinuxFlavor.RPM;
                }
            }
        }

        return LinuxFlavor.UNKNOWN;
    }

    private List<String> fetchInstalledApps(LinuxFlavor flavor) {
        String command;

        switch (flavor) {
        case DEBIAN:
            command = "dpkg-query -W -f='${Package}|${Version}|${Architecture}|${Installed-Size}|${Maintainer}|${Source}|${Homepage}\\n'";
            break;
        case RPM:
            command = "rpm -qa --queryformat '%{NAME}|%{VERSION}-%{RELEASE}|%{ARCH}|%{SIZE}|%{PACKAGER}|%{SOURCERPM}|%{URL}\\n'";
            break;
        default:
            LOG.warn("Installed Apps stats are not supported for this flavor of Linux OS");
            return Collections.emptyList();
        }

        return ExecutingCommand.runNative(command);
    }

    private List<LinuxAppInfo> parseApplicationInfo(List<String> output) {
        //logic to parse the output and store it in a list of LinuxAppInfo DTO
    }
}
```

External Programs can call the API as follows, below is the example from Mac

```
class Main {
  public static void main(String[] args) {
        MacInstalledApps installedApps = new MacInstalledApps();
        List<InstalledApps.MacAppInfo> macApps = installedApps.getInstalledApps();
        for (int i = 0; i < macApps.size(); i++) {
            System.out.println(macApps.get(i).toString());
        }
    }
}
```

**Testing:**

Executed `mvn test` and `oshi-core-6.7.1-SNAPSHOT.jar` got generated
Added it to my sample Java program and below are the outputs for the different OS

_Mac:_

```
MacAppInfo{appName='Automator', version='2.10', obtainedFrom='Apple', lastModified='15/11/24, 10:29', kind='Universal', signedBy='Software Signing',location='/System/Applications/Automator.app', getInfoString='unknown'}
MacAppInfo{appName='Books', version='7.1', obtainedFrom='Apple', lastModified='15/11/24, 10:29', kind='Universal', signedBy='Software Signing, location='/System/Applications/Books.app', getInfoString='unknown'}
...
```

_Windows:_

```
WindowsAppInfo{name=Notepad++ (64-bit x64), version=8.6.7, vendor=Notepad++ Team, installDate=, installLocation=, installSource=}
WindowsAppInfo{name=Microsoft Visual C++ 2019 X64 Minimum Runtime - 14.29.30153, version=14.29.30153, vendor=Microsoft Corporation, installDate=20240214, installLocation=, installSource=C:\ProgramData\Package Cache\{0F03096E-F81F-48D0-AEE0-9F85
13CD883F}v14.29.30153\packages\vcRuntimeMinimum_amd64\}
...
```

_Linux (Ubuntu):_

```
LinuxAppInfo{packageName='adduser', version='3.137ubuntu1', architecture='all', installedSize='452', maintainer='Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>', source='', homepage=''}
...
```

_Linux (Redhat):_

```
LinuxAppInfo{packageName='libgcc', version='11.5.0-2.el9', architecture='x86_64', installedSize='198756', maintainer='Red Hat, Inc. <http://bugzilla.redhat.com/bugzilla>', source='gcc-11.5.0-2.el9.src.rpm', homepage='http://gcc.gnu.org'}
...
```

Please let me know your views on the above.

Also, is the no. of variables shown in the case of Windows/Linux sufficient in the output.
If not, please let me know which ones to add.

**Comment by dbwiddis** [comment]
Hey @rohan-coder02 thanks so much for engaging on this!   It's really appreciated.

FYI, while I really appreciate the actual code implementation, there's no need to post a bunch of code in an issue, particularly for a complex feature like this.  You can always create a PR (you can mark it "Draft" if you're just looking for feedback) which gains a few benefits:
 - others are able to download your code and run it or otherwise evaluate it using more powerful IDE tools than eyeballs
 - it's easier to navigate between files
 - I can comment directly on a line of code or reference line numbers, etc., to make reviews a lot more succinct.
 - If you've already written the code it's probably easier to create a PR than to copy/paste all the code here! 

Some quick feedback on what you've posted so far:
 - I see some `List<?>` but we should have a defined interface that anything in that list implements.  In this case we probably want an `ApplicationInfo` interface
   - We can put getters for the most common/overlapping fields in this interface.  In other OSHI features I've generally used the WMI fields/names, so the Windows version seems closest to what I'd suggest here
   - We can include a getter for a key-value map that includes all available "optional" fields so we don't lose data
 - I see you're using `String` for pretty much everything.  Try to use types that don't require string parsing when possible (e.g., `long` for sizes and date-time stamps (in epoch milliseconds)

**Comment by dbwiddis** [comment]
Here's my thoughts on interface:

| Interface Field   | Windows         | macOS           | Linux           |
|-----------------|-----------------|-----------------|-----------------|
| name            | name            | appName         | packageName     |
| version         | version         | version         | version         |
| installLocation | installLocation | location        | ?             |
| installDate     | installDate     | lastModified    | ?             |
| vendor          | vendor          |  obtainedFrom (for Apple)<br />SignedBy (for Identified Developer)   | maintainer      |
| installSource   | installSource   | ?    | source          |

Additional OS-specific fields (to be stored in additionalInfo map):

| OS      | Additional Fields                    |
|---------|--------------------------------------|
| Windows | (Any extra fields not listed above)  |
| macOS   | kind, signedBy, getInfoString                  |
| Linux   | architecture, installedSize, homepage|

**Comment by rohan-coder02** [comment]
Hi @dbwiddis 

1)

Thanks for your feedback.
I've raised PR -> https://github.com/oshi/oshi/pull/2851

2)

I investigated further on installDate/lastModified in dpkg and rpm.
The following parameter helps (db-fsys:Last-Modified) in dpkg and (INSTALLTIME) in rpm.
I'm yet to find parameter for location as seen in windows/mac using a parameter like above.

**Comment by dbwiddis** [comment]
> I investigated further on installDate/lastModified in dpkg and rpm. The following parameter helps (db-fsys:Last-Modified) in dpkg and (INSTALLTIME) in rpm. I'm yet to find parameter for location as seen in windows/mac using a parameter like above.

How does location differ from source?

**Comment by dbwiddis** [comment]
Merged in @2851

**Comment by dyorgio** [comment]
Hi @dbwiddis.
I know that you don't has a Windows machine to test, but I believe that Windows implementation (Advapi32Util.registryGetStringValue) needs to use samDesiredExtra parameter to works when a Java 32 bits program runs in a Windows 64 program... (My guess is that needs two runs, one with WinNT.KEY_WOW64_32KEY and another with WinNT.KEY_WOW64_64KEY)
I suspect it because only few itens are returned on some machines that I tested it.
I'm studying about it (Windows API tricks), so I don't have a conclusion yet.
Maybe @rohan-coder02 or another Windows DEV could give us a light about this topic.

**Comment by dbwiddis** [comment]
> I suspect that using `WinNT.KEY_WOW64_64KEY` is enough. Maybe only when host machine is 64 bits.

According to [this SO answer](https://stackoverflow.com/a/12796797/1161484):

>  - KEY_WOW64_64KEY on a 64-bit OS means that the registry access, no matter if it's a 32 or 64 bit process, will access the 64 bit registry view.
>  - KEY_WOW64_32KEY on a 64-bit OS means that the registry access, no matter if it's a 32 or 64 bit process, will access the 32 bit registry view.
>  - Neither of them have any effect on a 32-bit OS.
>  - Leaving the flag out (the default) on a 64-bit OS will send registry accesses from 32-bit processes to the 32 bit registry view, and accesses from 64-bit processes to the 64 bit registry view.

This would imply it's safe to just always add that flag since it gives what we want on a 64-bit OS and has no effect on a 32-bit OS.

Although the last bullet implies leaving it out should give the same result?  Or not?

**Comment by rohan-coder02** [comment]
Hi @dyorgio and @dbwiddis 

I have raised PR for this issue.
https://github.com/oshi/oshi/pull/2902

I've added the analysis and replication details in the above PR.
